dtm0/doc/dtx.txt

```diff
@@ -0,0 +1,441 @@
+
+## dix_send
+
+This is a call that leads to posting of CAS-related RPC items.
+dix_rop posts them one by one. The state of each CAS request
+is being changed to SENT. Once such an even occur, it has to be
+delivered back to DTX immediately (through a callback), so that
+the state of the corresponding PA is moved to INPROGRESS.
+
+
+## cas_repl
+
+cas_repl is a group of the following calls:
+
+cas_req_replied_cb is called by RPC when it receives the reply.
+This sm transition is propagated to the upper DIXREQ through a clink connected
+to cas request SM.
+Then, DIXREQ propagates it up to DTX using the pointer to m0 client operation.
+DTX marks the corresponding PA state as "Executed", and updates
+the DTM0 log record.
+If any(dtx.pa == EXECUTED) then m0client calls op_executed callback
+and changes its state.
+
+## dtm_recv
+
+dtm_recv is a part of DTM service. The service should
+find the transaction (dtx) in the log and make a call.
+A dtx is linked with the corresponding m0 client op
+(container_of or parent-child pointers),
+so that DTX-related code should change the state of the operation
+once the condition all(dtx.pa == PERSISTENT) is satisfied.
+Also, it may move all dtx.pa[] into STABLE state.
+
+--------------------------------------------------------------------------------
+# DTX connections in psudo-code
+
+struct dtm_log {
+	struct record dl_records[];
+	struct rwlock dl_records_lock;
+};
+
+struct pa {
+	fid  pa_id; // DTM service id
+	bool pa_executed; // got reply
+	bool pa_persistent; // got PERSISTENT notice
+};
+
+struct txr {
+	struct pa t_pa[];
+	/// ... version, tx_id and so on
+};
+
+struct record {
+	struct txr    r_txr;
+	struct rwlock r_txr_lock;
+	struct dtx   *r_dtx;
+};
+
+/* dtx.d_sm states:
+	Init
+	Prepared
+	InProgress
+	Executed
+	Stable
+
+   dtx.d_sm transitions:
+	Init -> InProgress: all CAS requests are sent.
+	InProgress -> Executed: when got one reply (TBD).
+	Executed -> Stable: when txr is stable (TBD).
+*/
+
+struct dtx {
+	struct m0_sm        d_sm;
+	struct m0_sm_group *d_sm_group; // provided by M0 client
+	struct record      *d_record; // weak reference
+	struct m0_op       *d_op; // weak reference
+};
+
+dtx_invariant(dtx):
+	/* Assume: dtx and record are "cross-linked" */
+	ergo(dtx.d_record != NULL, dtx.d_record->d_dtx == dtx) AND
+	/* Assume: sm group is set when op is known */
+	ergo(dtx.d_op != NULL, dtx->d_sm_group != NULL)
+
+/* m0 Client operation */
+struct m0_op {
+	struct dix_req  op_dix_req;
+	// ...
+	struct dtx     *op_dtx;
+};
+
+m0op_invariant(op):
+	// Assume: when DTM0 flag is disabled,
+	//         op_dtx must be zeroed.
+	ergo(!DTM0, op->op_dtx == NULL)
+
+struct dix_req {
+	struct cas_req  dr_cas_req[];  {request,reply}
+	struct m0op    *dr_op;
+	struct dtx     *dr_dtx; // borrowed from op->op_dtx
+};
+
+
+dix_req_create(...) {
+	if (op->dtx == NULL) {
+		/* Preserve non-DTM logic */
+		clink_init(..., dix_req_clink_cb);
+	} else {
+		clink_init(..., dtx_clink_cb);
```

Comment:
Probably we'd need to be subscribed on both dix_req_clink_cb and dtx_clink_cb. JFYI.